lib/heuristics/scheduling_heuristic.rb

```diff
@@ -1008,47 +1004,70 @@ def get_stop(day, vrp, type, data = {})
         departure_time: data[:end],
         type: type,
         alternative: (data[:activity] if service_in_vrp&.activities),
-        detail: {
-          lat: associated_point.location&.lat,
-          lon: associated_point.location&.lon,
-          timewindows: ((service_in_vrp.activity&.timewindows || service_in_vrp.activities[data[:activity]].timewindows)&.select{ |t| t.day_index == day % 7 }&.collect{ |tw| { start: tw.start, end: tw.end } } if type == 'service'),
-          quantities: service_in_vrp&.quantities&.collect{ |qte| { unit: qte.unit.id, value: qte.value, label: qte.unit.label } } || [],
-          setup_duration: data[:considered_setup_duration],
-          duration: (data[:end] - data[:arrival] if type == 'service'),
-          skills: (service_in_vrp.skills if type == 'service')
-        }.delete_if{ |_k, v| !v }
+        detail: case type
+                when 'service'
+                  Wrappers::Wrapper.new.build_detail(
+                    service_in_vrp, service_in_vrp.activities[data[:activity]], associated_point,
+                    day % 7, nil, vehicle
+                  )
+                else
+                  Wrappers::Wrapper.new.build_detail(nil, nil, associated_point, nil, nil, vehicle)
+                end
       }.delete_if{ |_k, v| !v }
     end
 
-    def collect_route_stops(route_data, day, vrp)
+    def collect_route_stops(route_data, day, vrp, vehicle)
       previous = route_data[:start_point_id]
       route_data[:stops].collect{ |stop|
         stop[:travel_time] = matrix(route_data, previous, stop[:point_id])
         stop[:travel_distance] = matrix(route_data, previous, stop[:point_id], :distance)
         previous = stop[:point_id]
 
-        get_stop(day, vrp, 'service', stop)
+        get_stop(day, vrp, 'service', vehicle, stop)
       }
     end
 
+    def get_route_data(route_data)
+      route_start = route_data[:stops].empty? ? route_data[:tw_start] : route_data[:stops].first[:start]
+
+      route_end, final_travel_time, final_travel_distance =
+        if route_data[:stops].empty?
```

Comment:
It seems that some parts could be calculated apart instead of inside a multiple level if condition

Reply:
I am afraid it will make three different functions (one for route_end, one for final_travel_time, one for final_travel_distance) with same if/elsif/else. 

What do you have in mind ?